src/V3Increments.cpp

```diff
@@ -0,0 +1,222 @@
+//*************************************************************************
+// DESCRIPTION: Verilator: Replace return/continue with jumps
+//
+// Code available from: https://verilator.org
+//
+//*************************************************************************
+//
+// Copyright 2003-2020 by Wilson Snyder. This program is free software; you
+// can redistribute it and/or modify it under the terms of either the GNU
+// Lesser General Public License Version 3 or the Perl Artistic License
+// Version 2.0.
+// SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0
+//
+//*************************************************************************
+// V3Increments's Transformations:
+//
+// Expand each pre and post incrementation/decrementation into separate
+// statements.
+//
+//*************************************************************************
+
+#include "config_build.h"
+#include "verilatedos.h"
+
+#include "V3Global.h"
+#include "V3Increments.h"
+#include "V3Ast.h"
+
+#include <algorithm>
+#include <cstdarg>
+#include <vector>
+
+#include <iostream>
+
+//######################################################################
+
+class IncrementsVisitor : public AstNVisitor {
+private:
+    // TYPES
+    enum InsertMode {
+        IM_BEFORE,  // Pointing at statement ref is in, insert before this
+        IM_AFTER,  // Pointing at last inserted stmt, insert after
+        IM_WHILE_PRECOND  // Pointing to for loop, add to body end
+    };
+
+    // STATE
+    int m_modIncrementsNum;  // Var name counter
+    InsertMode m_insMode;  // How to insert
+    AstNode* m_insStmtp;  // Where to insert statement
+    bool m_unsupportedHere;  // Used to detect where it's not supported yet
+
+    // METHODS
+    void insertBefore(AstNode* placep, AstNode* newp) {
+        AstNRelinker linker;
+        placep->unlinkFrBack(&linker);
+        newp->addNext(placep);
+        linker.relink(newp);
+    }
+
+private:
+    void insertBeforeStmt(AstNode* nodep, AstNode* newp) {
+        // Return node that must be visited, if any
+        // See also AstNode::addBeforeStmt; this predates that function
+        if (debug() >= 9) newp->dumpTree(cout, "-newstmt:");
+        UASSERT_OBJ(m_insStmtp, nodep, "Function not underneath a statement");
+        if (m_insMode == IM_BEFORE) {
+            // Add the whole thing before insertAt
+            if (debug() >= 9) newp->dumpTree(cout, "-newfunc:");
+            m_insStmtp->addHereThisAsNext(newp);
+        } else if (m_insMode == IM_AFTER) {
+            m_insStmtp->addNextHere(newp);
+        } else if (m_insMode == IM_WHILE_PRECOND) {
+            AstWhile* whilep = VN_CAST(m_insStmtp, While);
+            UASSERT_OBJ(whilep, nodep, "Insert should be under WHILE");
+            whilep->addPrecondsp(newp);
+        } else {
+            nodep->v3fatalSrc("Unknown InsertMode");
+        }
+        m_insMode = IM_AFTER;
+        m_insStmtp = newp;
+    }
+
+    virtual void visit(AstWhile* nodep) VL_OVERRIDE {
```

Comment:
Since there's special code for while (e.g. expanded FORs) your test should make sure to try the four operators in all four loop locations (pre, test, post) body.  Unless other tests well cover this.